dissect/target/plugins/apps/container/docker.py

```diff
@@ -48,128 +113,162 @@ class DockerPlugin(Plugin):
+                        ports=convert_ports(ports),
+                        names=config.get("Name").replace("/", "", 1),
+                        volumes=volumes,
+                        source=fp,
+                        _target=self.target,
+                    )
+
+    @export(record=DockerLogRecord)
+    @arg(
+        "--raw-messages",
+        action="store_true",
+        help="preserve ANSI escape sequences and trailing newlines from log messages",
+    )
+    @arg(
+        "--remove-backspaces",
+        action="store_true",
+        help="alter messages by removing ASCII backspaces and the corresponding characters",
+    )
+    def logs(self, raw_messages: bool = False, remove_backspaces: bool = False) -> Iterator[DockerLogRecord]:
+        """Returns log files (stdout/stderr) from Docker containers.
+
+        The default Docker Daemon log driver is ``json-file``, which
+        performs no log rotation. Another log driver is ``local`` and
+        performs log rotation and compresses log files more efficiently.
+
+        Eventually ``local`` will likely replace ``json-file`` as the
+        default log driver.
+
+        Resources:
+            - https://docs.docker.com/config/containers/logging/configure/
+            - https://docs.docker.com/config/containers/logging/json-file/
+            - https://docs.docker.com/config/containers/logging/local/
+        """
+
+        for data_root in self.installs:
+            containers_path = f"{data_root}/containers"
+
+            for container in self.target.fs.path(containers_path).iterdir():
+                # json-file log driver
+                for log_file in container.glob(f"{container.name}-json.log*"):
+                    for line in open_decompress(log_file, "rt"):
+                        try:
+                            log_entry = json.loads(line)
+                        except json.JSONDecodeError as e:
+                            self.target.log.warning(f"Could not decode JSON line in file {log_file}")
+                            self.target.log.debug("", exc_info=e)
+                            continue
+
+                        yield DockerLogRecord(
+                            ts=log_entry.get("time"),
+                            container=container.name,
+                            stream=log_entry.get("stream"),
+                            message=log_entry.get("log")
+                            if raw_messages
+                            else strip_log(log_entry.get("log"), remove_backspaces),
+                            _target=self.target,
+                        )
+
+                # local log driver
+                for log_file in container.glob("local-logs/container.log*"):
+                    fh = open_decompress(log_file, "rb")
+                    pos = 0
+
+                    if not hasattr(fh, "size"):  # for pytest
+                        fh.size = len(fh.read())
+                        fh.seek(0)
```

Comment:
If this is only for pytest. Can you fix it in pytest itself?

Reply:
You will notice I left this in the suggested code as well (albeit some changes). It should still be fixed in pytest itself there.

Reply:
I attempted to mock this in pytest, which is a headache. What do you suggest we do here? I think it would be benificial to have the `size` attribute available just as you would outside of pytest. 

Do you suggest adding `size` to `MappedFile` or `TargetPath`? I am too unfamiliar with these classes to decide which is a good solution.

Reply:
I don't think `.size` is available either in the case of a compressed file, so this would break then as well.

I think the best approach to do here is just run the file out until it's EOF. That way you also don't unnecessarily decompress until the end of the file in case of a compressed file. This is a common pattern we use with cstruct.

```python
while True:
    try:
        entry = c_local.entry(fh)
        ...
    except EOFError:
        break
```